github/VisualStudio
# Expose API to show GitHub Login Dialog

- Depends on #2403

## What this PR does

Add a simple API for showing the GitHub login dialog. This API is exposed as a MEF service, so that it can be easily consumed by Team Explorer.

```cs
namespace GitHub.VisualStudio.Essentials
{
    public interface ILoginService
    {
        void ShowLoginDialog();
    }
}
```

Since the GitHub Essentials extension has dependencies on Team Explorer, it might make sense for the `ILoginService` interface to be exposed by Team Explorer and implemented/exported by GitHub Essentials.

Consuming this API could be prototyped as follows without adding any new Team Explorer interfaces.

```cs
    [Export]
    public class ConsumeLoginService
    {
        readonly dynamic loginService;

        [ImportingConstructor]
        public ConsumeLoginService([Import("GitHub.VisualStudio.Essentials.ILoginService", AllowDefault = true)] dynamic loginService)
        {
            this.loginService = loginService;
        }

        void SomethingThatRequiresGitHubAuthentication()
        {
            if (loginService is null)
            {
                return; // Login service not available
            }

            loginService.ShowLoginDialog();
        }
    }
```

## How to test

1. Clone https://github.com/github/visualstudio
2. Open `GitHubVS.sln`
3. Right-click `Restore NuGet Packages`
4. Build solution
5. `Solution Explorer > Folder View`
6. Open `src\GitHub.VisualStudio.16.sln`
7. Rename `...\Common7\IDE\Extensions\GitHub\GitHub.16\extension.vsixmanifest` to `extension.vsixmanifest`
8. Rename `...\Common7\IDE\Extensions\GitHub\GitHub.16\GitHub.VisualStudio.16.pkgdef` to `GitHub.VisualStudio.16.pkgdef_`
9. Build solution
10. Try something like `ConsumeLoginService` above

If everything goes to plan, you should see the `Connect to GitHub` dialog:

![image](https://user-images.githubusercontent.com/11719160/64982062-3ae0c500-d8b5-11e9-8e5b-c66887c0ea64.png)
